aten/src/ATen/native/TensorFactories.cpp

```diff
@@ -727,5 +727,195 @@ Tensor tensor_cuda(ArrayRef<T> values, const TensorOptions& options) {
   }
 AT_FORALL_SCALAR_TYPES_EXCEPT_HALF(TENSOR)
 #undef TENSOR
+
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ choice ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+template <typename scalar_t>
+void generate_keys(
+  scalar_t *keys,
+  scalar_t *weights,
+  int n,
+  THGenerator* generator
+){
+  std::lock_guard<std::mutex> lock(generator->mutex);
+
+  for(int i = 0; i < n; i++){
+    scalar_t u = THRandom_standard_uniform(generator);
+    keys[i] = weights[i] > 0? (scalar_t) std::pow(u, 1/weights[i]):-1;
+  }
+
+}
+
+void reservoir_generator_cpu(
+  int64_t *indices,
+  int64_t n,
+  int64_t k,
+  THGenerator* generator
+){
+  std::lock_guard<std::mutex> lock(generator->mutex);
+
+  for(int i = k; i < n; i++){
+    int64_t z = THRandom_random(generator) % (i + 1);
+    if (z < k) {
+        std::swap(indices[z], indices[i]);
+    }
+  }
+
+}
+
+Tensor reservoir_sampling_cpu(
+  const Tensor& x,
+  const Tensor& weights,
+  int64_t k
+){
+
+  AT_CHECK(
+    weights.dtype() == kFloat,
+    "The sampling weights must be Float, got", weights.dtype()
+  );
+
+  AT_CHECK(
+    weights.is_contiguous(),
+    "The sampling weights must be contiguous."
```

Comment:
why?

Reply:
You are right, It doesn't. I can just check whether the weights are contiguous and in the case they aren't I can just force it to be. I'll make this change, thanks.